### ISSUE TITLE
Warnings and false-positives using type aliases in an object-like

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#911]
```php
<?php
/**
 * @psalm-type aType null|'a'|'b'|'c'|'d'
 */

/** @psalm-return array{0:bool,1:aType} */
function f(): array
{
    return [(bool)rand(0,1), rand(0,1) ? 'z' : null];
}
```
```
Scanning files...
PHP Notice:  Undefined offset: 17 in /home/weirdan/.composer/vendor/vimeo/psalm/src/Psalm/Type.php on line 595

Notice: Undefined offset: 17 in /home/weirdan/.composer/vendor/vimeo/psalm/src/Psalm/Type.php on line 595
Analyzing files...

ERROR: InvalidReturnType - q.php:6:5 - The declared return type 'array{0:bool, 1:string|null}' for f is incorrect, got 'array{0:bool, 1:string|null}'
/** @psalm-return array{0:bool,1:aType} */
function f(): array
{
    return [(bool)rand(0,1), rand(0,1) ? 'z' : null];
}


ERROR: InvalidReturnStatement - q.php:9:5 - The type 'array{0:bool, 1:string|null}' does not match the declared return type 'array{0:bool, 1:string|null}' for f
    return [(bool)rand(0,1), rand(0,1) ? 'z' : null];


------------------------------
2 errors found
------------------------------

Checks took 0.09 seconds and used 24.960MB of memory
```
Psalm dev-master@5ca35a0635234e8bce82a6bcf14512191e70bf33
